### DISCUSSION TITLE
How does tcppktlat tool work ?

### DISCUSSION THREAD

**Discussion by 0xFA32** [iovisor/bcc#4908] [General]
As per the documentation of tcppktlat tool, it seems to trace the latency between TCP received pkt and picked up by userspace thread. 

And looking at the implementation it seems to use the raw tracepoint  `tcp_probe` and `tcp_rcv_space_adjust` to track the incoming and outgoing time to then calculate the latency.

`tcp_probe` tracepoint is at `tcp_rcv_established` which to my understanding is when the packet is received when the connection is in an established state. Which makes sense to me to track the incoming time.

However, I don't quite get how using `tcp_rcv_space_adjust` would provide us the outgoing time.. Isn't it traced when the receive buffer needs to be adjusted, which isn't always. So wouldn't it be better if `tcp_recvmsg` is used ?

**Comment by yonghong-song** [comment]
@mind-fuzzy The current tcppktlat can only track incoming packet latency between tcp receiving packet and the packet assigned to a user space. it does not support tracking latency for outgoing packets.

**Reply by 0xFA32** [reply]
> the packet assigned to a user space

ah yes, I was wondering if `tcp_rcv_space_adjust` provides the right tracepoint to track when the packet is assigned to user space.

**Reply by 0xFA32** [reply]
> This function should be called every time data is copied to user space. It calculates the appropriate TCP receive buffer space.

My bad. https://elixir.bootlin.com/linux/v6.1.14/source/net/ipv4/tcp_input.c#L701
